src/pretix/plugins/paypal/views.py

```diff
@@ -31,36 +31,71 @@
 # Unless required by applicable law or agreed to in writing, software distributed under the Apache License 2.0 is
 # distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 # License for the specific language governing permissions and limitations under the License.
-
+import hashlib
 import json
 import logging
-from decimal import Decimal
 
-import paypalrestsdk
-import paypalrestsdk.exceptions
 from django.contrib import messages
 from django.core import signing
-from django.db.models import Sum
-from django.http import HttpResponse, HttpResponseBadRequest
+from django.http import (
+    Http404, HttpResponse, HttpResponseBadRequest, JsonResponse,
+)
 from django.shortcuts import get_object_or_404, redirect, render
 from django.urls import reverse
+from django.utils.decorators import method_decorator
 from django.utils.translation import gettext_lazy as _
 from django.views.decorators.clickjacking import xframe_options_exempt
 from django.views.decorators.csrf import csrf_exempt
 from django.views.decorators.http import require_POST
+from django.views.generic import TemplateView
 from django_scopes import scopes_disabled
-from paypalrestsdk.openid_connect import Tokeninfo
+from paypalcheckoutsdk.orders import OrdersGetRequest
+from paypalhttp import HttpError
 
-from pretix.base.models import Event, Order, OrderPayment, OrderRefund, Quota
+from pretix.base.models import Event, Order, OrderPayment, Quota
 from pretix.base.payment import PaymentException
+from pretix.base.settings import GlobalSettingsObject
 from pretix.control.permissions import event_permission_required
 from pretix.multidomain.urlreverse import eventreverse
+from pretix.plugins.paypal.client.customer.partners_merchantintegrations_get_request import (
+    PartnersMerchantIntegrationsGetRequest,
+)
 from pretix.plugins.paypal.models import ReferencedPayPalObject
-from pretix.plugins.paypal.payment import Paypal
+from pretix.plugins.paypal.payment import PaypalMethod, PaypalMethod as Paypal
+from pretix.presale.views import get_cart, get_cart_total
 
 logger = logging.getLogger('pretix.plugins.paypal')
 
 
+class PaypalOrderView:
+    def dispatch(self, request, *args, **kwargs):
+        try:
+            self.order = request.event.orders.get(code=kwargs['order'])
+            if hashlib.sha1(self.order.secret.lower().encode()).hexdigest() != kwargs['hash'].lower():
+                raise Http404('Unknown order')
+        except Order.DoesNotExist:
+            # Do a hash comparison as well to harden timing attacks
+            if 'abcdefghijklmnopq'.lower() == hashlib.sha1('abcdefghijklmnopq'.encode()).hexdigest():
+                raise Http404('Unknown order')
+            else:
+                raise Http404('Unknown order')
+        return super().dispatch(request, *args, **kwargs)
+
+    @property
```

Comment:
we usually use a @cached_property for properties like this unless we really want a database query every time we use it

Reply:
Copy/Pasted from other pprov - seems like we are a little bit lax about this and have both `@property` and `@cached_property` in concurrent use across multiple payment providers.